### PR TITLE
Add all supported currencies for Amazon Pay

### LINKED ISSUE #4798: Update supported currencies for Amazon Pay
Per https://docs.stripe.com/payments/amazon-pay, Amazon Pay supports the following currencies: USD, AUD, GBP, DKK, EUR, HKD, JPY, NZD, NOK, ZAR, SEK, and CHF.

At the moment, we only allow USD for Amazon Pay -- we need to update the set of supported currencies to reflect all the available options.

### PR DESCRIPTION
**Pull request by daledupreez**
Fixes STRIPE-813
Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/4798

## Changes proposed in this Pull Request:

This PR updates Amazon Pay to correctly reflect all the currently supported countries and currencies. In particular, the following account countries can enable Amazon Pay: AT, BE, CY, DK, FR, DE, HU, IE, IT, LU, NL, PT, ES, SE, CH, GB, US

US accounts only support USD as the currency. Accounts from all other supported countries support all of the following currencies: AUD, CHF, DKK, EUR, GBP, HKD, JPY, NOK, NZD, SEK, USD, ZAR.

At present, there are two notable sets of edge cases:
 * There are a number of European countries using the Euro that aren't supported, including FI, GR, HR, LT, LV, SI, and SK.
 * There are a number of supported currencies where accounts in those countries aren't supported, including AUD, HKD, JPY, NZD, and ZAR.

Beyond the baseline logic in the payment method to support those definitions, I've also updated the client-side logic for supported currencies, and I've updated the express checkout helper to include country and currency checks for Amazon Pay.

There was also a fix for address normalization, as Amazon Pay test transactions in EUR only include address line 2 and leave address line 1 empty. I don't know if that will be true for production as well, but I added it in to allow for more complete testing.

Note that this PR does not update the settings UI to better handle unsupported currencies. We are tracking that separately in STRIPE-806 / https://github.com/woocommerce/woocommerce-gateway-stripe/issues/4782, as it requires some additional UI work and refactoring.

## Testing instructions

* Check out this branch and build the updated UI - `git checkout add/amazon-pay-all-supported-currencies`
* Set up a store with a US address and USD as the currency
* Disconnect from Stripe and connect to Stripe again
* Ensure that you can enable Amazon Pay successfully, and enable Amazon Pay for express checkout
* As a shopper, ensure you can complete a purchase using Amazon Pay
* Now change the store currency to a non-USD currency
* Ensure that we no longer show Amazon Pay as an express checkout option
   - You can also verify that `wc_stripe_express_checkout_params?.stripe?.is_amazon_pay_enabled` is `false` via your Javascript console in the product page, cart, or checkout.
* Disconnect your store from Stripe
* Update the store address to another supported country and currency, e.g. `BE` + `EUR` or `GB` + `GBP`
   - It would be good to pick EUR plus a European country to test the address normalization piece, as those flow through `amazon.de` and use a test German address with address line 2 populated but address line 1 empty.
* Reconnect to Stripe
* Ensure that Amazon Pay is still enabled -- it was likely disabled as a result of the reconnection
* Confirm that Amazon Pay is shown for purchases, and confirm that you can make a purchase using Amazon Pay
* Now change the store currency to another supported currency
* Reload the product page, cart, or checkout, and confirm that Amazon Pay is still visible
* Make a purchase and confirm that it flows through successfully
* Disconnect from Stripe
* Change the store country to be an unsupported country with a supported currency, e.g. `AU` + `AUD` or `GR` + `EUR`
* Reconnect to Stripe
* Confirm that Amazon Pay is not shown as an express checkout option in the payment settings UI (for now -- the method should be disabled, but we're going to handle that in a dedicated PR)
* Confirm that Amazon Pay is not shown to shoppers either

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [N/A] Tested on mobile (or does not apply)

### Changelog entry

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 1]
```diff
@@ -1696,7 +1702,18 @@ public function is_payment_request_enabled() {
 	public function is_amazon_pay_enabled() {
 		$is_enabled = WC_Stripe_UPE_Payment_Method_Amazon_Pay::is_amazon_pay_enabled( $this->gateway );
 
-		return $is_enabled && $this->is_enabled_for_current_context( 'amazon_pay' );
+		if ( ! $is_enabled || ! $this->is_enabled_for_current_context( 'amazon_pay' ) ) {
+			return false;
+		}
+		$amazon_pay_instance = new WC_Stripe_UPE_Payment_Method_Amazon_Pay();
```
Comment (Mayisha) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php:1717]: We are creating a new instance everytime `is_amazon_pay_enabled` called. This may lead to issues like multiple hooks registration, multiple cache/db queries in the parent constructor. 
I vaguely remember we once had issues with multiple subscription hook registration due to multiple instances of the gateway.

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 2]
```diff
@@ -1319,6 +1319,12 @@ public function fix_address_fields_mapping( $data ) {
 			}
 		}
 
+		// Amazon Pay may allow address data in address_2 instead of address_1.
+		if ( empty( $data['billing_address']['address_1'] ) && ! empty( $data['billing_address']['address_2'] ) ) {
+			$data['billing_address']['address_1'] = wc_clean( wp_unslash( $data['billing_address']['address_2'] ) );
+			$data['billing_address']['address_2'] = '';
+		}
```
[the remark this answers is not in the document]
Comment (Mayisha) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:1331]: It should be safe to check this too.

**Review by diegocurbelo — APPROVED** [review]
The code looks good, and it tests as described.

I found a small error with the testing instructions, the PR description mentions `account countries`, which is correct because we use `is_available_for_account_country()` to check availability, but the test instructions mention changing the store address country and reconnecting, which worked for me only when reconnecting with different Stripe accounts (US, FR, PL)

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php (hunk 1 above)
Comment (daledupreez) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:1717]: I just checked the inheritance, and I don't believe this is an issue. The class in question doesn't perform anything other than property initialisation, and the parent class is `WC_Stripe_UPE_Payment_Method`, which also limits itself to local property initialisation _and_ doesn't call the parent constructor. Instantiating a new gateway instance would definitely cause problems, but I think we're OK instantiating only the payment method instance.

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (daledupreez) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:1331]: Done in 14589f2a30cc4561819b079c3e5d077270af3993.

**Review by Mayisha — APPROVED** [review]
Looks good to me. The availability of the Amazon Pay express checkout button is working as described.

✅ Account: US, Currency: USD -> Amazon Pay present in settings and storefront.
✅ Account: US, Currency: Euro -> Amazon Pay present in settings with currency pill and absent from storefront.
✅ Account: AU, Currency: AUD -> Amazon Pay absent from settings and storefront.
✅ Account: GB, Currency: AUD -> Amazon Pay present in settings and storefront.
 
 **Note**: I was unable to test a transaction as Amazon account creation is failing for me.
